Shorten error descriptions

ColonelGerdauf opened the pull request:
This will reduce the verbal clutter when an update has several mods which throw error(s). Keeping the errors to one line will allow room for more errors within the precious default textbox size.

ColonelGerdauf committed:
Shorten error descriptions (CF)

Signed-off-by: ColonelGerdauf <ColonelGerdauf@users.noreply.github.com>

ColonelGerdauf committed:
Shorten error descriptions (MR)

Signed-off-by: ColonelGerdauf <ColonelGerdauf@users.noreply.github.com>

Kationor commented on launcher/modplatform/flame/FlameCheckUpdate.cpp:
```diff
@@ -140,15 +140,14 @@ void FlameCheckUpdate::executeTask()
             auto pack = getProjectInfo(latest_ver.value());
             auto recover_url = QString("%1/download/%2").arg(pack.websiteUrl, latest_ver->fileId.toString());
-            emit checkFailed(mod, tr("Mod has a new update available, but is not downloadable using CurseForge."), recover_url);
+            emit checkFailed(mod, tr("Mod update is not downloadable using Prism."), recover_url);
```
```suggestion
            emit checkFailed(mod, tr("Cannot download mod update due to CurseForge restrictions."), recover_url);
```
Slightly longer, but infinitely more useful than the previous error message.

ColonelGerdauf replied:
This would imply that there is a systemic restriction from CurseForge, but it is something that specific mod authors have enabled as a setting. If I recall correctly, this was the default setting at one point.

Most mods you are able to download just fine, but there are a few where the setting may or may not have been applied intentionally.

ColonelGerdauf replied:
What I can suggest is "Cannot download update due to author restrictions." Would this be a better fit for what you intend?

ColonelGerdauf replied:
I have submitted the changes requested.

BeansBeefBroccoli commented:
I don't have any authority or anything but I think this is a bad change

ColonelGerdauf committed:
Improve description

Based on feedback on PR#2853

Signed-off-by: ColonelGerdauf <ColonelGerdauf@users.noreply.github.com>

getchoo commented on launcher/modplatform/flame/FlameCheckUpdate.cpp:
```diff
@@ -140,15 +140,14 @@ void FlameCheckUpdate::executeTask()
         setStatus(tr("Parsing the API response from CurseForge for '%1'...").arg(mod->name()));
 
         if (!latest_ver.has_value() || !latest_ver->addonId.isValid()) {
-            emit checkFailed(mod, tr("No valid version found for this mod. It's probably unavailable for the current game "
```
I'm not sure how this is "verbal clutter". It's adding pretty important context to this error that will (hopefully) help users understand why the update is failing in most cases

ColonelGerdauf replied:
Saying that there is no valid version is pretty straightforward. 

The latter sentence is unnecessary, especially with how vague it is; it does not in fact provide users with useful information. "probably unavailable" is more confusing than it is helpful, and the implication is already clear with the stated inability to find a valid path and therefore lacks a valid ID.

In order for additional information to be useful, it needs to be fairly specific, though maybe not to the point that each warning is special. As it stands, all that extra fluff does is make people impulsively press "Next" or "Cancel", which then lowers the awareness of why a mod update has failed.

This to me is a prime case of why KISS is so important in the bigger picture.

getchoo replied:
> Saying that there is no valid version is pretty straightforward.

Yes, but it fails to explain *why* there might not be a valid version

> The latter sentence is unnecessary, especially with how vague it is

"game *version*" would probably be a better clarification, but I don't see this as grounds to remove it

> "probably unavailable" is more confusing than it is helpful

And "No valid version found" is not? I think most would question why that is at first. Regardless, "probably" doesn't make things more confusing IMO, it just offers another solution

> and the implication is already clear with the stated inability to find a valid path and therefore lacks a valid ID.

I have no idea how "No valid version found for this mod" makes any of this "clear". You specifically calling it an implication should be a good sign of this

> In order for additional information to be useful, it needs to be fairly specific

Is something like "it may not support the current version of the game" not fairly specific?
I would also say the base "No valid version found for this mod" is extremely vague

> As it stands, all that extra fluff does is make people impulsively press "Next" or "Cancel"

How?

> which then lowers the awareness of why a mod update has failed.

And removing the most common reason as to why no valid version can be found does...? A lot of what you're saying here is frankly arguing against your changes. You want things to be very specific and help users understand why a mod update failed, but you're removing information about why that might be. Only saying "No valid version found for this mod" is extremely vague and does not do anything to help understand the cause;  it only describes the issue

 > This to me is a prime case of why KISS is so important in the bigger picture.
 
If KISS means removing 7 words explaining why something failed 99% of the time just for the sake of shortening it, I would rather not use this principle for writing error message 🫤

ColonelGerdauf replied:
You are describing the latter sentence as if it is important, useful, and vital. It may be so in the back-end and for those who are already familiar with the structures of Prism's coding. News flash: the vast majority are not. 

Let me ask: how exactly does the latter sentence explain in any meaningful way why Prism could not find a valid version? Implicit information goes a very long way. Saying that "there is no valid version that Prism could find" tells a lot; it says to the person that the mod is either locked, or removed, or is in a different MC version, or any random thing. In any way, the mod is not available for download, and that is often sufficient for when a user runs into difficulties.

Saying " It's probably unavailable for the current game..." gives zero clarifying information. You are telling people "okay, just maybe, the mod never existed". This is the opposite of helpful. You want a helpful second sentence? Try specifying; "The mod is available for a different version" or "The mod does not appear to exist". You are saying probably... probably what, exactly? This is the ambiguity that can eventually lead people to a lot of problematic assumptions about mod author or Prism dev intentions.

Yes, I am removing information: information that is obscenely vague and will only serve to annoy people and ignore the meat of the problem in the first place. You want the cause explained? Thus far, I have not seen a particularly good effort in that direction, and let's not forget one of the driving factors of the change: the textbox size that displays the errors is a bit small, and having a better understanding of which mods and how many are having issues right off the bat is much more useful than knowing "more about the problem", which let me repeat, is not what that word salad is accomplishing.

The key principle of KISS is to write things including error messages in a way that 95% of people can understand the gist quickly, not that 2% or so can understand the problem inside and out. This sounds like a classic Linux fallacy, and I for one will not stand for it. In a lot of situations, information overload is worse to deal with than being short-changed on the specifics, especially down to the consumer/client level.

getchoo replied:
> You are describing the latter sentence as if it is important, useful, and vital

It is useful, and can give important context to why the error is happening

> It may be so in the back-end and for those who are already familiar with the structures of Prism's coding.

Why is it only useful in the back-end? Nothing internal to Prism is described, there's no technical jargon; it's explaining that there is no version of the mod available for the current game version. This requires 0 knowledge of "the structures of Prism's coding", only knowing that mods on sites like Modrinth and CF only work for specific versions of the game. I imagine most people downloading mods are aware of this

> Let me ask: how exactly does the latter sentence explain in any meaningful way why Prism could not find a valid version? 

I will repeat my previous comment: "game *version*" would probably be a better clarification, but I don't see this as grounds to remove it

> Yes, I am removing information: information that is obscenely vague and will only serve to annoy people and ignore the meat of the problem in the first place

Why not improve it as opposed to removing it? Again, repeating my previous comment: You want things to be very specific and help users understand why a mod update failed, but you're removing information about why that might be.

> You want the cause explained? Thus far, I have not seen a particularly good effort in that direction

I already suggested a change that could better clarify the meaning behind this message. I also don't see why you're bringing this up when you're explicitly removing it and not putting a "good effort in that direction" either

>  and let's not forget one of the driving factors of the change: the textbox size that displays the errors is a bit small, and having a better understanding of which mods and how many are having issues right off the bat is much more useful than knowing "more about the problem"

Maybe it would be, but that's not what this PR is doing. It's removing 7 words and nothing else. Having the actual reason behind an error at least be *somewhere* is also very useful in communicating to the user *why* it's happening. Having a vague message shown over and over again with no reason as to why is just annoying

> which let me repeat, is not what that word salad is accomplishing.

How would you suggest we fix this "word salad"

> The key principle of KISS is to write things including error messages in a way that 95% of people can understand the gist quickly, not that 2% or so can understand the problem inside and out.

Ok? You can still quickly and easily explain what the error is and why it's most likely happening. "No valid version found for this mod" is extremely vague, and adding a few more Simple words can resolve that

> This sounds like a classic Linux fallacy, and I for one will not stand for it.

Briefly explaining why an error occurs is a "classic Linux fallacy"? Would you rather we follow the Windows method of just showing a random error code and saying "Something went wrong"?

>  In a lot of situations, information overload is worse to deal with than being short-changed on the specifics, especially down to the consumer/client level.

I honestly feel like we're reading completely different things here. If a few extra words is really information overload to you, I don't think there is much point in continuing this conversation. Feel free to suggest improvements, but I will not be approving a pull request that "fixes" vague error messages by removing more from them and compounding the issue

ColonelGerdauf replied:
For clarification, what I am referring to when I say "classic Linux fallacy" is to put a specific format and magically expecting everyone to follow suit, even though, a lot of the time in this context, it is really detached from the actual userbase outside the 7331 coders, and is therefore completely useless to them.

This is an issue that has plagued projects such as GIMP. Now, why do Inkscape, Krita, or Blender not share the same fate (for the most part)? Because the developers have actively listened to not only the wider userbase, but also industry feedback. They understand what they are dealing with, not in terms of competition, but more so in terms of what the users want to have and the limits of what they can actually tolerate.

Only if the FOSS CAD devs could do the same...

Why I cited the fallacy is to point to the scale of who can understand the information, versus who will simply skip it.

If, according to you, "a few simple words" is enough to render the message useless, then I would like to ask what you have as an alternative. Is there a if/else somewhere that can append to the error string? If not, then the idea that the edit offered will make things "more vague" makes zero sense. It is still the same type of error, and people can read the implications from a simplified sentence.

Remember, the information needs to be succinct enough to fit in a single line in the default Prism error textbox, which is the larger point of this PR.

getchoo replied:
> For clarification, what I am referring to when I say "classic Linux fallacy" is to put a specific format and magically expecting everyone to follow suit, even though, a lot of the time in this context, it is really detached from the actual userbase outside the 7331 coders, and is therefore completely useless to them.

This is a lot of hyperbole. Once again, "7331 coders" are not the only ones who can understand this or get help from it. Most people attempting to update a mod that isn't compatible with their game version yet will be greatly assisted by an error message telling them exactly that

> This is an issue that has plagued projects such as GIMP. Now, why do Inkscape, Krita, or Blender not share the same fate (for the most part)? Because the developers have actively listed to not only the wider userbase, but also industry feedback. They understand what they are dealing with, not in terms of competition, but more so in terms of what the users want to have and the limits of what they can actually tolerate.
> 
> Only if the FOSS CAD devs could do the same...

At this point we're getting very...very off topic, and frankly grasping at straws

> Why I cited the fallacy is to point to the scale of who can understand the information, versus who will simply skip it.

I think there's a good middleground here. I am well aware a good chunk of users will skip over messages like these, but that doesn't mean we should not provide any information to those who take the time to read it and possibly fix their issue. If we were to just ignore those people, what's the point in clarifying anything in an error message? Again: Would you rather we follow the Windows method of just showing a random error code and saying "Something went wrong"?

> I would like to ask what you have as an alternative. Is there a if/else somewhere that can append to the error string?
 
I think there may be some miscommunication here. This is now the third time I have suggested this, so to make it abundantly clear this time: I think something along the lines of ***"It may be unavailable for the current game version"*** is a better way of getting the point of this message across

> It is still the same type of error, and people can read the implications from a simplified sentence.

"Saying 'No valid version found for this mod' gives zero clarifying information. You are telling people 'okay, just maybe, the mod never existed'. This is the opposite of helpful"

> Remember, the information needs to be succinct enough to fit in a single line in the default Prism error textbox, which is the larger point of this PR.

Why? If the "larger point of this PR" is to only ensure error messages fit into a completely arbitrary limit of one line and not actually improve them, it's extremely misguided


In any case, this conversation has continued to go on in circles. I've made my opinion clear and given suggestions multiple times; you have made your views clear as well. I would like for this to continue, but until these suggestions are taken into account, there isn't much point in doing so. This will be my last comment until that happens

ColonelGerdauf replied:
I have attempted to provide information and make the description magically "less vague", but what I am unhappy with is the fact that the "more information" came at the cost of spotty specificity. It tends to be an either or; you either provide all the situations plainly (log file), or you only give enough for contextual implications (user-facing error messages).

Guess what limitation I am dealing with in this PR...

getchoo commented on launcher/modplatform/modrinth/ModrinthCheckUpdate.cpp:
```diff
@@ -214,7 +214,7 @@ void ModrinthCheckUpdate::checkNextLoader()
     }
     for (auto m : m_mappings) {
         emit checkFailed(m,
-                         tr("No valid version found for this mod. It's probably unavailable for the current game version / mod loader."));
```
Ditto

ColonelGerdauf replied:
See above change request

getchoo commented on launcher/modplatform/flame/FlameCheckUpdate.cpp:
```diff
@@ -140,15 +140,14 @@ void FlameCheckUpdate::executeTask()
             auto pack = getProjectInfo(latest_ver.value());
             auto recover_url = QString("%1/download/%2").arg(pack.websiteUrl, latest_ver->fileId.toString());
-            emit checkFailed(mod, tr("Mod has a new update available, but is not downloadable using CurseForge."), recover_url);
+            emit checkFailed(mod, tr("Cannot download update due to author restrictions."), recover_url);
```
```suggestion
            emit checkFailed(mod, tr("Unable to update mod due to author restrictions."), recover_url);
```

Keeping context in this error message doesn't make much sense after removing the above -- though I'm definitley glad it's still here 😄 

I do wonder though: are "author restrictions" really the only reason standard API downloads from Curseforge would be blocked? In the case they aren't, I would rather keep the original message here as this could be misleading

Ryex replied:
The are only three cases where a download link for a mod update would be missing that I can think of

1) author has blocked 3rd party clients for the mod that it was previously enabled for
2) author or curseforge has pulled the the upload for this version
3) the mod has updated but the new version doesn't support this game version (pretty sure we filter this out before )

ColonelGerdauf replied:
I actually like the suggestion. It is verbally shorter, and sounds a bit less in-your-face.

And my concern with the original "Curseforge restrictions" is what I had explained previously; it incorrectly implies that it is Curseforge itself that is preventing any mod from being updated externally. The pointed phrasing there can induce panic and confusion, especially with how inconsistent it would sound. While there was a bit of a scare in that direction, after concessions it is a setting used by users, either intentionally, or an infrequent mod author who was unaware of automatic "defaults" that were there at one point of time.

ColonelGerdauf committed:
Update launcher/modplatform/flame/FlameCheckUpdate.cpp

Noice

Co-authored-by: seth <getchoo@tuta.io>
Signed-off-by: ColonelGerdauf <ColonelGerdauf@users.noreply.github.com>

ColonelGerdauf committed:
Attempt at simplification w/ contextual information

If this does not satisfy the "problem" at hand, then I am lost as to what to do.

Signed-off-by: ColonelGerdauf <ColonelGerdauf@users.noreply.github.com>

ColonelGerdauf committed:
Attempt at simplification w/ contextual information

If this does not satisfy the "problem" at hand, then I am lost as to what to do.

Signed-off-by: ColonelGerdauf <ColonelGerdauf@users.noreply.github.com>

TheKodeToad commented:
To be honest the discussion here doesn't look very productive. I think it is better that this PR is closed or only the change which Kationor made to be kept (perhaps with "most likely" added though).

If the vagueness/uncertainty is an issue - "It's *probably* unavailable for the current game version / mod loader"... code complexity could be introduced to make the error box actually check that it is the reason... but while I do like some attention to detail that doesn't feel too important (and I think that might involve much more code changes than it's worth).

I don't think just taking out potentially useful information does anything good for the user.

ColonelGerdauf commented:
As I've been trying to explain, useful information is not the end all be all; enough information needs to be provided, sure, but there is also the matter of attention span, in relation to situations where the error shows up, as well as density of immediately important data. IE how many mods are effected with errors, rather than all the wild possibilities of what is happening with each mod.

The error detail has a maximum of 80 characters in the default box before the text jumps to the next line. When there is only so many types of error messages, and they are all lengthy, this robs people of knowing the bigger picture of what mods and how many are affected, which then prompts many to skip the message prompt and potentially overlook information that could have been helpful.

Even without the matter of character length in boxes, there comes a point where the "useful" error messages will come across as gibberish, especially when repeated many times over. Diversity of errors is a different matter that I hoped would be approached after this type of PR suggestion. There is maybe four error messages to catch tens to hundreds of different situations; this honestly flies in the face of the information being "useful", especially when taking into consideration long-term usage.

ColonelGerdauf committed:
I give up...

This is all I am left with, taking to account details that were never properly explained to me in the first place.

Signed-off-by: ColonelGerdauf <ColonelGerdauf@users.noreply.github.com>

ColonelGerdauf committed:
I give up...

This is all I am left with, taking to account details that were never properly explained to me in the first place.

Signed-off-by: ColonelGerdauf <ColonelGerdauf@users.noreply.github.com>

TheKodeToad commented:
At this point the amount of discussion about some text in the UI feels a little excessive; I feel like this is not worth it
